tools/template_based_query_generation/src/main/java/Node.java

```diff
@@ -0,0 +1,124 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.Random;
+
+/**
+ * class representing node for a query in the markov chain
+ */
+public class Node<E> {
+
+    private E obj;
+    private HashMap<Node<E>, Double> neighbors;
+    private ArrayList<Node<E>> neighborList; // list of neighbors and corresponding cumulative probabilities
+    private ArrayList<Double> cProbabilities; // cumulative probabilities
+    private Random r;
```

Comment:
Should we keep a Random instance in each node or all of the nodes should share the same instance?

Reply:
All of them share the same instance, since they are initialized in the same class with the seed given by the user. Is it a problem if they share the same instance? Would they be generating similar random numbers then? Is it a problem if they have different instances?

Reply:
With the current implementation I believe it each not will have its own instance as the instance is initialized in the contructor?
Also, vote +1 for sharing the same instance. If they do not share the same instance, I **guess** with the same seed each node will get same list of random numbers.

Reply:
When the Nodes are created in QueryGenerator, they will be given the same Random r. I have added a constructor in Node to take in an instance of Random.